src/vs/workbench/contrib/notebook/browser/notebookExecutionStateServiceImpl.ts

```diff
@@ -142,14 +143,87 @@ export class NotebookExecutionStateService extends Disposable implements INotebo
 		return exe;
 	}
 
-	private _setLastFailedCell(notebook: URI, cellHandle: number) {
-		this._lastFailedCells.set(notebook, cellHandle);
-		this._onDidChangeLastRunFailState.fire({ failed: true, notebook });
+	private _setLastFailedCell(notebookURI: URI, cellHandle: number) {
+		const prevLastFailedCellInfo = this._lastFailedCells.get(notebookURI);
+		const newLastFailedCellInfo: IFailedCellInfo = { cellHandle: cellHandle, visible: true };
+		if (prevLastFailedCellInfo) {
+			newLastFailedCellInfo.disposable = prevLastFailedCellInfo.disposable;
+		} else {
+			newLastFailedCellInfo.disposable = this._getFailedCellListener(notebookURI);
+		}
+		this._lastFailedCells.set(notebookURI, newLastFailedCellInfo);
+
+		this._onDidChangeLastRunFailState.fire({ visible: true, notebook: notebookURI });
+	}
+
+	private _setLastFailedCellVisibility(notebookURI: URI, visible: boolean) {
+		const lastFailedCellInfo = this._lastFailedCells.get(notebookURI);
+
+		if (lastFailedCellInfo) {
+			this._lastFailedCells.set(notebookURI, {
+				cellHandle: lastFailedCellInfo.cellHandle,
+				disposable: lastFailedCellInfo.disposable,
+				visible: visible,
+			});
+		}
+
+		this._onDidChangeLastRunFailState.fire({ visible: visible, notebook: notebookURI });
+	}
+
+	private _clearLastFailedCell(notebookURI: URI) {
+
+		const lastFailedCellInfo = this._lastFailedCells.get(notebookURI);
+
+		if (lastFailedCellInfo) {
+			lastFailedCellInfo.disposable?.dispose();
+			this._lastFailedCells.delete(notebookURI);
+		}
+
+		this._onDidChangeLastRunFailState.fire({ visible: false, notebook: notebookURI });
 	}
 
-	private _clearLastFailedCell(notebook: URI) {
-		this._lastFailedCells.delete(notebook);
-		this._onDidChangeLastRunFailState.fire({ failed: false, notebook: notebook });
+	private _getFailedCellListener(notebookURI: URI) {
```

Comment:
Not everyone does this, but it's a good idea to always include the return types for methods in the signature, even when it's void. That would make it more apparent that this method returns `IDisposable | undefined`, which is a little odd, is it possible to get here with a notebook URI that does not exist?